deltacat/logs.py

```diff
@@ -19,6 +24,45 @@
 DEFAULT_DEBUG_DELTACAT_LOG_BASE_FILE_NAME = "deltacat-python.debug.log"
 
 
+class RayRuntimeContextLoggerAdapter(logging.LoggerAdapter):
+    """
+    Logger Adapter for injecting Ray Runtime Context into logging messages.
+    """
+    def __init__(self, logger: Logger, runtime_context: RuntimeContext):
+        super().__init__(logger, {})
+        self.runtime_context = runtime_context
+
+    def process(self, msg, kwargs):
+        """
+        Injects Ray Runtime Context details into each log message.
+
+        This may include information such as the raylet node ID, task/actor ID, job ID,
+        placement group ID of the worker, and assigned resources to the task/actor.
+
+        Args:
+            msg: The original log message
+            kwargs: Keyword arguments for the log message
+
+        Returns: A log message with Ray Runtime Context details
+
+        """
+        runtime_context_dict = self.runtime_context.get()
+        if self.runtime_context.get_task_id() or self.runtime_context.get_actor_id():
+            runtime_context_dict["pg_id"] = self.runtime_context.get_placement_group_id()
+            runtime_context_dict["assigned_resources"] = self.runtime_context.get_assigned_resources()
+
+        return "(ray_runtime_context=%s) -- %s" % (runtime_context_dict, msg), kwargs
```

Comment:
```
2023-02-15 06:16:08,111	INFO pid=1777 dedupe.py:168 -- (ray_runtime_context={'job_id': JobID(01000000), 'node_id': NodeID(93fe7173180d4b441a970c1a0a40c5c7f17d8870daa365e008cc5590), 'namespace': '3b627167-2f9f-4294-9696-785d43e2ffcb', 'task_id': TaskID(a816bd897febbb3fffffffffffffffffffffffff01000000), 'pg_id': None, 'assigned_resources': {'CPU': 1.0, 'node:172.31.35.247': 0.0001}}) -- [Dedupe task 104] Getting delta file envelope groups for 188 object refs...
```
Would it make more sense to suffix the `runtime_context` data to the log message for readability as opposed to injecting it before the rest of LogRecord attributes like `%msg`

Something like this 
```
2023-02-15 06:16:08,111	INFO pid=1777 dedupe.py:168 -- [Dedupe task 104] Getting delta file envelope groups for 188 object refs... -- runtime_context={'job_id': JobID(01000000), 'node_id': NodeID(93fe7173180d4b441a970c1a0a40c5c7f17d8870daa365e008cc5590), 'namespace': '3b627167-2f9f-4294-9696-785d43e2ffcb', 'task_id': TaskID(a816bd897febbb3fffffffffffffffffffffffff01000000), 'pg_id': None, 'assigned_resources': {'CPU': 1.0, 'node:172.31.35.247': 0.0001}})
```

Reply:
By convention, logger context is prefixed to the log message. For example, date and log level is also part of the context. I am not aware of exact reason as to why but at the same time I don't believe suffixing it would make any difference to readability.

Reply:
> ```
> 2023-02-15 06:16:08,111	INFO pid=1777 dedupe.py:168 -- (ray_runtime_context={'job_id': JobID(01000000), 'node_id': NodeID(93fe7173180d4b441a970c1a0a40c5c7f17d8870daa365e008cc5590), 'namespace': '3b627167-2f9f-4294-9696-785d43e2ffcb', 'task_id': TaskID(a816bd897febbb3fffffffffffffffffffffffff01000000), 'pg_id': None, 'assigned_resources': {'CPU': 1.0, 'node:172.31.35.247': 0.0001}}) -- [Dedupe task 104] Getting delta file envelope groups for 188 object refs...
> ```
> 
> Would it make more sense to suffix the `runtime_context` data to the log message for readability as opposed to injecting it before the rest of the immediately relevant LogRecord attributes like `%msg`
> 
> Something like this
> 
> ```
> 2023-02-15 06:16:08,111	INFO pid=1777 dedupe.py:168 -- [Dedupe task 104] Getting delta file envelope groups for 188 object refs... -- runtime_context={'job_id': JobID(01000000), 'node_id': NodeID(93fe7173180d4b441a970c1a0a40c5c7f17d8870daa365e008cc5590), 'namespace': '3b627167-2f9f-4294-9696-785d43e2ffcb', 'task_id': TaskID(a816bd897febbb3fffffffffffffffffffffffff01000000), 'pg_id': None, 'assigned_resources': {'CPU': 1.0, 'node:172.31.35.247': 0.0001}})
> ```

I've tried something like that as well. Readability is subjective, but what I didn't like about the log message sandwiched between two contexts is that it now becomes hard to just eyeball the actual log message you're looking for. Having the log message always be at the tail end of each log makes them a bit easier to dissect, even with all the bloated context.